### PR TITLE
Kd make solver

### PR DESCRIPTION
**Pull request by kdomino**

### CODE REVIEW

File: src/PEPS.jl [hunk 1]
```diff
@@ -82,59 +25,116 @@ mutable struct PepsNetwork
...
+MPO(R::PEPSRow) = MPO(Float64, R)
+
+function MPO(::Type{T},
+    peps::PepsNetwork,
+    i::Int,
+    config::Dict{Int, Int} = Dict{Int, Int}()
+    ) where {T <: Number}
+
+    W = MPO(T, peps.j_max)
+    for (j, A) ∈ enumerate(PEPSRow(peps, i))
```
Comment (lpawela) [inline src/PEPS.jl]: This should call `PEPSRow(T, peps, i)`

File: src/PEPS.jl [hunk 2]
```diff
@@ -82,59 +25,116 @@ mutable struct PepsNetwork
...
+        W[j] = B
+    end
+    W
+end
+MPO(peps::PepsNetwork,
+    i::Int,
+    config::Dict{Int, Int} = Dict{Int, Int}()
+    ) = MPO(Float64, peps, i, config)
+
+    function boundaryMPS(
```
Comment (lpawela) [inline src/PEPS.jl]: why this is indented?

File: src/PEPS.jl [hunk 3]
```diff
@@ -82,59 +25,116 @@ mutable struct PepsNetwork
...
+                ψ = compress(ψ, Dcut, tol, max_sweeps)
+            end
+            push!(vec, ψ)
+        end
+        if reversed reverse(vec) else vec end
+    end
 
-        @cast A[_, u, _, d] |= exp(-ng.β * en[u, d]) 
-        ψ[j] = A
+function LightGraphs.contract(
```
Comment (lpawela) [inline src/PEPS.jl:114]: This does make sense in the context of the original `LightGraphs.contract`

File: src/base.jl [hunk 4]
```diff
@@ -49,6 +49,16 @@ end
 @inline MPS(A::AbstractArray, ::Val{:right}, Dcut::Int, args...) = _left_sweep_SVD(A, Dcut, args...)
 @inline MPS(A::AbstractArray, ::Val{:left}, Dcut::Int, args...) = _right_sweep_SVD(A, Dcut, args...)
 
+@inline Base.dropdims(ψ::MPS, i::Int) = (dropdims(A, dims=i) for A ∈ ψ)
+@inline Base.dropdims(ψ::MPS) = Base.dropdims(ψ, 2)
+
+function idMPS(::Type{T}, L::Int) where {T <: Number}
```
Comment (lpawela) [inline src/base.jl:55]: How is this used?

File: src/network.jl [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+function generate_tensor(ng::NetworkGraph, v::Int)
+    fg = ng.factor_graph
+    loc_exp = exp.(-ng.β .* get_prop(fg, v, :loc_en))
+
+    dim = []
+    @cast tensor[_, i] := loc_exp[i]
+
+    for w ∈ ng.nbrs[v]
+        if has_edge(fg, w, v)
```
Comment (lpawela) [inline src/network.jl:32]: I still don't line this `if` statement. There should be a nicer way to implement this

File: test/PEPS.jl [hunk 6]
```diff
@@ -47,157 +47,16 @@ x, y = m, n
...
-
-    for A ∈ ψ @test size(A, 2) == 1 end
-
-    for i ∈ 2:peps.i_max
-        println("row -> ", i)
-
-        R = PEPSRow(peps, i)
-        W = MPO(R)
-        M = MPO(peps, i-1, i)
+    ψ = idMPS(peps.j_max)
```
Comment (lpawela) [inline test/PEPS.jl:53]: If `idMPS` is needed for testing, please do not implement it in the package. We should not put methods needed only in tests in the main code.

File: src/utils.jl [hunk 7]
```diff
@@ -213,8 +214,17 @@ end
     end
 end
 
-function rank_vec(ig::MetaGraph)
```
Comment (lpawela) [inline src/utils.jl:216]: Why is this removed?

File: test/ising.jl [hunk 8]
```diff
@@ -3,6 +3,34 @@ using LightGraphs
 using GraphPlot
 using CSV
 
+function _energy(config::Dict, couplings::Dict, cedges::Dict, n::Int)
+    eng = zeros(1,n)
+    for (i, j) ∈ keys(cedges)
```
Comment (lpawela) [inline test/ising.jl:8]: Please put this monstrosity in a single `for`

File: src/PEPS.jl (hunk 1 above)
Comment (bartekGardas) [reply src/PEPS.jl]: Indeed!

File: src/PEPS.jl (hunk 2 above)
Comment (bartekGardas) [reply src/PEPS.jl]: The same convention as in e.g., factor_graph.

File: src/PEPS.jl (hunk 3 above)
Comment (bartekGardas) [reply src/PEPS.jl:114]: In a sense, yes it does since both functions contract a graph.

File: src/base.jl (hunk 4 above)
Comment (bartekGardas) [reply src/base.jl:55]: Thus far, in `contrac`t and `boundaryMPS`, but will be in more places. Ideally, it would be nice to have something like `MPS(I)` or maybe we should initialize `MPS` with ones. Then calling `MPS(T, L)` would be enough.  I think that both are valid solutions.

File: src/network.jl (hunk 5 above)
Comment (bartekGardas) [reply src/network.jl:32]: As the network can be general (not only square), I do not believe we can avoid` if` here.

File: test/PEPS.jl (hunk 6 above)
Comment (bartekGardas) [reply test/PEPS.jl:53]: This is not only for testing, we need the identity MPS for sure.

File: test/ising.jl (hunk 8 above)
Comment (bartekGardas) [reply test/ising.jl:8]: This test (and many others) will be rewritten by @dexter2206. Before that happen, I would leave it as is.

File: src/utils.jl (hunk 7 above)
Comment (bartekGardas) [reply src/utils.jl:216]: How come all tests pass if this function is removed?

File: test/PEPS.jl (hunk 6 above)
Comment (lpawela) [reply test/PEPS.jl:53]: Ok, I'll introduce an `IdentityMPS` and `IdentityMPO` types after this is merged.

File: benchmarks/low_energy_approx_chimera.jl [hunk 9]
```diff
@@ -71,13 +71,25 @@ problem_size = parse_args(s)["size"]
...
-
-ig = ising_graph(fi, si, 1)
+spectrum_cutoff = parse_args(s)["spectrum_cutoff"]
 
 n_sols = parse_args(s)["n_sols"]
 node_size = (parse_args(s)["node_size1"], parse_args(s)["node_size2"])
 println(node_size)
 
+
+s1 = Int(sqrt(si/8))
```
Comment (lpawela) [inline benchmarks/low_energy_approx_chimera.jl]: use `isqrt`

File: benchmarks/low_energy_approx_chimera.jl [hunk 10]
```diff
@@ -71,13 +71,25 @@ problem_size = parse_args(s)["size"]
...
+n = ceil(Int, s1/node_size[1])
```
Comment (lpawela) [inline benchmarks/low_energy_approx_chimera.jl]: use `div` or `÷`

File: benchmarks/low_energy_approx_chimera.jl [hunk 11]
```diff
@@ -86,9 +98,8 @@ ground_ref = [parse(Int, el) for el in data[i][4:end]]
 ground_spins = binary2spins(ground_ref)
 energy_ref = energy(ground_spins, ig)
 
-spectrum_cutoff = parse_args(s)["spectrum_cutoff"]
 
-ses = collect(spectrum_cutoff:-10:40)
+ses = collect(spectrum_cutoff:-1:1)
```
Comment (lpawela) [inline benchmarks/low_energy_approx_chimera.jl]: do we need `collect` here?

File: benchmarks/low_energy_approx_chimera.jl [hunk 12]
```diff
@@ -86,9 +98,8 @@ ground_ref = [parse(Int, el) for el in data[i][4:end]]
...
 step = 10
 n_s = collect(n_sols:-step:1)
```
Comment (lpawela) [inline benchmarks/low_energy_approx_chimera.jl]: again, why `collect`?

File: src/utils.jl (hunk 7 above)
Comment (lpawela) [reply src/utils.jl:216]: It is in `ising.jl`

File: src/PEPS.jl (hunk 1 above)
Comment (lpawela) [reply src/PEPS.jl]: Any update on this?

File: src/PEPS.jl (hunk 2 above)
Comment (lpawela) [reply src/PEPS.jl]: Please put this on the same identation level as the rest of the code

File: src/base.jl (hunk 4 above)
Comment (lpawela) [reply src/base.jl:55]: I'll add something like this after the merge

File: src/PEPS.jl (hunk 3 above)
Comment (lpawela) [reply src/PEPS.jl:114]: Maybe we should use `solve`?

File: src/PEPS.jl (hunk 1 above)
Comment (kdomino) [reply src/PEPS.jl]: done

File: benchmarks/low_energy_approx_chimera.jl (hunk 9 above)
Comment (kdomino) [reply benchmarks/low_energy_approx_chimera.jl]: done

File: benchmarks/low_energy_approx_chimera.jl (hunk 10 above)
Comment (kdomino) [reply benchmarks/low_energy_approx_chimera.jl]: done

File: benchmarks/low_energy_approx_chimera.jl (hunk 12 above)
Comment (kdomino) [reply benchmarks/low_energy_approx_chimera.jl]: done

File: benchmarks/low_energy_approx_chimera.jl (hunk 11 above)
Comment (kdomino) [reply benchmarks/low_energy_approx_chimera.jl]: done

File: src/PEPS.jl (hunk 2 above)
Comment (kdomino) [reply src/PEPS.jl]: done

File: src/PEPS.jl (hunk 3 above)
Comment (bartekGardas) [reply src/PEPS.jl:114]: I think `contract` is better as this is exactly what happens. Note, this function does not solve for the ground state, just contract the network for a given configuration or assumes lack thereof.

Comment (kdomino) [reply src/PEPS.jl:114]: I agree, solve would overwrite existing function

Comment (bartekGardas) [reply src/PEPS.jl:114]: Overwriting is not an issue here. It is just contract does not involve actual solving, only simple contraction.

**Review by lpawela — APPROVED** [review]
